demos/coreMQTT/mqtt_demo_mutual_auth.c

```diff
@@ -670,19 +644,21 @@ int RunCoreMqttMutualAuthDemo( bool awsIotMqttMode,
 
 static int32_t prvGenerateRandomNumber()
 {
-    const uint32_t ulMultiplier = 0x015a4e35UL, ulIncrement = 1UL;
-
-    ulNextRand = ( ulMultiplier * ulNextRand ) + ulIncrement;
-    return( ( int32_t ) ( ulNextRand & 0x7ffffffUL );
-}
+    uint32_t ulRandomNum;
 
-/*-----------------------------------------------------------*/
+    /* Use the PKCS11 module to generate a random number. */
+    if( xPkcs11GenerateRandomNumber( ( uint8_t * ) &ulRandomNum,
+                                     ( sizeof( ulRandomNum ) == pdPASS ) ) )
+    {
+        ulRandomNum = ( ulRandomNum & INT32_MAX );
```

Comment:
This is probably okay, but it assumes that the upper and lower halves of the distribution are completely independent and it throws away a bit of randomness.  This is the reason that `xPkcs11GenerateRandomNumber()` returns a _status_ which is independent from the _value_ returned BTW.  Any time status and value are combined in a return value, you're forced to make tradeoffs like this.  Here, we're eliminating 2B worth of possible random values just so that we can have a single negative return value.

The best way to solve this class of problem is to preseve the range of values.  Does `prvGenerateRandomNumber()` even need to exist, or could this demo simply call `xPkcs11GenerateRandomNumber()`?  That would be simpler and it would preserve all bits of randomness.

Reply:
This wrapper function is required for the RNG interface type in `FreeRTOS/backoffAlgorithm` which is defined as `typedef int32_t ( * BackoffAlgorithm_RNG_t )();`

As the backoff algorithm library only generates a backoff value that is 16 bits wide, losing the MSB worth of randomness in a 32 bit wide integer is fine (imo) here as the library ultimately normalizes the random value down to the max jitter window which cannot be more than 16 bits wide..